File: internal/models/command.go
package models

import "gorm.io/gorm"

// Command represents a system command that can be executed
type Command struct {
	gorm.Model
	Name        string `gorm:"uniqueIndex;not null"`
	Description string
	Command     string             `gorm:"not null"`
	Category    string             `gorm:"index"`
	Enabled     bool               `gorm:"default:true"`
	Parameters  []CommandParameter `gorm:"foreignKey:CommandID"`
	Executions  []CommandExecution `gorm:"foreignKey:CommandID"`
}

// ParameterType represents the type of a command parameter
type ParameterType string

const (
	ParameterTypeString  ParameterType = "string"
	ParameterTypeNumber  ParameterType = "number"
	ParameterTypeBoolean ParameterType = "boolean"
	ParameterTypeSelect  ParameterType = "select"
	ParameterTypeFile    ParameterType = "file"
)

// CommandParameter represents parameters for a command
type CommandParameter struct {
	gorm.Model
	CommandID    uint          `gorm:"index"`
	Name         string        `gorm:"not null"`
	Label        string        `gorm:"not null"`
	Type         ParameterType `gorm:"not null"`
	Required     bool          `gorm:"default:false"`
	DefaultValue string
	Options      string // JSON string for select options
	Validation   string // Regex or validation rules
	Order        int    `gorm:"default:0"`
}

// CommandExecution represents a record of command execution
type CommandExecution struct {
	gorm.Model
	CommandID    uint   `gorm:"index"`
	UserID       uint   `gorm:"index"`
	Parameters   string // JSON string of parameters
	Status       string `gorm:"not null"` // success, error, etc.
	Result       string // Output of the command
	ErrorMessage string
	Duration     int64 // Duration in milliseconds
	IPAddress    string
}

File: frontend/src/models/command.ts
// Command represents a system command that can be executed
export interface Command {
    ID: number;
    createdAt: string;
    updatedAt: string;
    deletedAt?: string | null;
    name: string;
    description?: string;
    command: string;
    category?: string;
    enabled: boolean;
    parameters: CommandParameter[];
    executions: CommandExecution[];
}

// ParameterType represents the type of a command parameter
export type ParameterType =
    | "string"
    | "number"
    | "boolean"
    | "select"
    | "file";

// CommandParameter represents parameters for a command
export interface CommandParameter {
    id: number;
    createdAt: string;
    updatedAt: string;
    deletedAt?: string | null;
    commandId: number;
    name: string;
    label: string;
    type: ParameterType;
    required: boolean;
    defaultValue?: string;
    options?: string; // JSON string for select options
    validation?: string; // Regex or validation rules
    order: number;
}

// CommandExecution represents a record of command execution
export interface CommandExecution {
    id: number;
    createdAt: string;
    updatedAt: string;
    deletedAt?: string | null;
    commandId: number;
    userId: number;
    parameters: string; // JSON string of parameters
    status: string; // success, error, etc.
    result?: string; // Output of the command
    errorMessage?: string;
    duration: number; // Duration in milliseconds
    ipAddress?: string;
}
